### PR TITLE
Add `bit_view` support class

### PR DESCRIPTION
**Pull request by kcgen**
The `bit_view` class is a wrapper around a data value. It provides a view into a subset of its bits allowing them to be read, written, assigned, flipped, cleared, and tested, without the need to for the usual twiddling operations.

Constructing a `bit_view` is similar to C bitfields, however unlike C bitfields, `bit_view`s are free from undefined behavior and have been proven using GCC's and Clang's undefined behavior sanitizers.

This gives us the benefit of bitfields without their specification downsides:
 - they're succinct and clear to use
 - they're just as fast as bitwise operators (maybe more so, being `constexpr`)
 - each `bit_view` is self-documenting by specifying its bit index, number of bits, and logical name
 - the view's range-of-bits are _locked-in_ using explicit template values, so there's no risk of them being "slippery" or changing during runtime.
 - the view's internal data member is always safely sized to accommodate the view's position and size. static-asserts ensure that it isn't under or over-specified.

### Before and after comparison

Assume we have a 16-bit audio control register that holds the card's:
 - left volume at bit 1 using 6 bits
 - right volume at bit 7, also 6 bits
 - speaker on/off state at bit 13 with just one bit

The existing DOSBox code would ferret out this information using manual bit twiddling:

``` c++
if (data & 0x2000)
	enable_speaker();
else
	disable_speaker();

const auto left_percent = ((data & 0x7e) >> 1) / 63.0;
const auto right_percent = ((data & 0x1f80) >> 7) / 63.0;
set_volume(left_percent, right_percent);
```

The `bit_view` makes the register's logical elements self-documenting:

``` c++
union AudioReg {
   uint16_t data = 0;
   bit_view<1, 6> left_volume;
   bit_view<7, 6> right_volume;
   bit_view<13, 1> speaker_on;
};
```

The views can be used directly, making for much clearer (and safer) code:

``` c++
AudioReg reg = {data};

if (reg.speaker_on)
	enable_speaker();
else
	disable_speaker();

const auto left_percent = reg.left_volume / 63.0;
const auto right_percent = reg.right_volume / 63.0;
set_volume(left_percent, right_percent);
```

### CODE REVIEW

**Comment by kcgen** [comment]
Todo:
 - assert sums to catch over/under flow
 - assert right-hand side type `is_integral` and not floating point
 - add unit tests that traffic `bit_view` objects for everyday use in function arguments, aliased as a reference, pointed to by pointers, newable, smart-,pointer-able, and used in fixed arrays.

**Comment by kcgen** [comment]
Similar concepts have been tried. For example, this [approach](https://preshing.com/20150324/safe-bitfields-in-cpp/) uses lots of preprocessor macros to build up a union and classes:

``` c++
BEGIN_BITFIELD_TYPE(Status, uint32_t) 
    ADD_BITFIELD_MEMBER(readers, 0, 10)  
    ADD_BITFIELD_MEMBER(waitToRead, 10, 10)
    ADD_BITFIELD_MEMBER(writers, 20, 10)
END_BITFIELD_TYPE()
```

To me, this no longer feels like clean easy to read C++ and more like a giant contraption. It surely gets the job done, but it's too much of a departure from easy-to-read C++ without any macro-hacks.

There's also this approach that uses `type::displacement`:

``` c++
union manually_done_bitfield {
    using element_t = long;
    using base_t = base<long, 4, 3, 5>;
    BitFieldMember<element_t, base_t::displacement(0), 4> fourBits;
    BitFieldMember<element_t, base_t::displacement(1), 3> threeBits;
    BitFieldMember<element_t, base_t::displacement(2), 5> fiveBits;
};
```

Plus some macros:

``` C++
// Generic macros
#define UNTUPLE_2_1(a, b) a
#define UNTUPLE_2_2(a, b) b
#define PP_SEQ_TUPLE_2_2(s, d, element) UNTUPLE_2_2 element
#define PP_SEQ_ENUM_UNTUPLE_2_2(r, data, e) ,UNTUPLE_2_2 e

// Macro to declare a member
#define PP_SEQ_I_MAKE_BITFIELD(r, data, i, element) BitFieldMember<element_t, base_t::displacement(i), UNTUPLE_2_2 element> UNTUPLE_2_1 element;
```

Which is feeling a bit more like standard C++:

``` c++
    BitFieldMember<element_t, base_t::displacement(0), 4> fourBits;
    BitFieldMember<element_t, base_t::displacement(1), 3> threeBits;
    BitFieldMember<element_t, base_t::displacement(2), 5> fiveBits;
```

However, these `displacement` values feel more like a second layer of meta-programming (given we're already using templates), and their values represent the position in the `base<long, 4, 3, 5>` as opposed to their starting bit positions of the view. 

I think the approach is transparent and simple enough that even a non-assembly guy like me can understand it, and the tests bear out that it's working without any undefined behaviour.

**Comment by kcgen** [comment]
###UBSAN tests

Build:

``` shell
meson setup -Dbuildtype=debug -Db_sanitize=undefined build/clang-undefined
meson compile -C build/clang-undefined
```

Results:

``` text
[03:06:13@server dosbox-staging] (kc/bit_view-1 $) ./build/clang-undefined/tests/bit_view
Running main() from gmock_main.cc
[==========] Running 18 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 18 tests from bit_view
[ RUN      ] bit_view.assign_from_literal
[       OK ] bit_view.assign_from_literal (0 ms)
[ RUN      ] bit_view.assign_to_data
[       OK ] bit_view.assign_to_data (0 ms)
[ RUN      ] bit_view.assign_from_parts
[       OK ] bit_view.assign_from_parts (0 ms)
[ RUN      ] bit_view.flip
[       OK ] bit_view.flip (0 ms)
[ RUN      ] bit_view.increment
[       OK ] bit_view.increment (0 ms)
[ RUN      ] bit_view.decrement
[       OK ] bit_view.decrement (0 ms)
[ RUN      ] bit_view.clear
[       OK ] bit_view.clear (0 ms)
[ RUN      ] bit_view.boolean_checks
[       OK ] bit_view.boolean_checks (0 ms)
[ RUN      ] bit_view.equality
[       OK ] bit_view.equality (0 ms)
[ RUN      ] bit_view.writable_via_reference
[       OK ] bit_view.writable_via_reference (0 ms)
[ RUN      ] bit_view.readable_via_reference
[       OK ] bit_view.readable_via_reference (0 ms)
[ RUN      ] bit_view.writable_via_pointer
[       OK ] bit_view.writable_via_pointer (0 ms)
[ RUN      ] bit_view.readable_via_pointer
[       OK ] bit_view.readable_via_pointer (0 ms)
[ RUN      ] bit_view.create_with_new
[       OK ] bit_view.create_with_new (0 ms)
[ RUN      ] bit_view.create_with_make_unique
[       OK ] bit_view.create_with_make_unique (0 ms)
[ RUN      ] bit_view.use_in_array
[       OK ] bit_view.use_in_array (0 ms)
[ RUN      ] bit_view.bare_initialization
[       OK ] bit_view.bare_initialization (0 ms)
[ RUN      ] bit_view.bare_constructor
[       OK ] bit_view.bare_constructor (0 ms)
[----------] 18 tests from bit_view (1 ms total)

[----------] Global test environment tear-down
[==========] 18 tests from 1 test suite ran. (1 ms total)
[  PASSED  ] 18 tests.
```

**Comment by shermp** [comment]
Looking nice.

From an ergonomics point of view, is there a way of creating the bit_view without explicitly setting the type? I know C++ can do some magic regarding templates and type inference, but unsure of when/how it can do so.

**Comment by kcgen** [comment]
@shermp , yeah, the position and width of the bit_view can give you the largest type needed for that specific bit_view. 

So that means, say, for a 16bit register, the first handful of bit_views would use uint8_t and the latter handful would use a uint16_t. 

The mixed-sizing made me nervous enough to try to avoid it, and make sure all the bit_views are aliasing the the same data-value.  (But that's perhaps unecessary concern!)

Ideally we could as for the size of the parent structure, and figure out that we're sitting inside a union 16-bits in size, then everything could use that. Something like that would get it to the 100% level.

It also eliminates any possibility of someone using the wrong type (although the static_asserts still make sure they'll never go out of range).

**Comment by johnnovak** [comment]
> The mixed-sizing made me nervous enough to try to avoid it, and make sure all the bit_views are aliasing /holding the same data-type. (But that's perhaps an unecessary concern!)

We're using the bit-views for low-level stuff only, and in that context the register/data widths are fixed and known in advance in 99.99% of the cases. So I very much prefer the current approach of having to specify the width explicitly for each register or data type we're dealing with; that's much more descriptive, reflects actual reality, and is a lot less error prone.

File: .clang-format
Comment (johnnovak) [inline .clang-format:70]: 👍🏻

**Comment by kcgen** [comment]
If we can lock'em in:

``` c++
union MyReg {
    uint8_t data; // this is the type we want to learn
    bit_view<...> // about and lock-in for all the bit_view
};
```
I've been scanning the web but can't find how to get the data type from the parent union/class (without nasty macro). 

Might have to settle for manually repeating it?

**Comment by johnnovak** [comment]
Nice addition, this will make all that low-level bit-twiddling less error-prone and more readable! 👍🏻 

> Might have to settle for manually repeating it?

If there was a simple way to avoid the manual repetition that would be preferable, of course, but I don't think it's worth complicating the code with lots of template/macro magic just to make that happen someho

**Review by johnnovak — APPROVED** [review]

File: tests/bit_view_tests.cpp
Comment (johnnovak) [inline tests/bit_view_tests.cpp:33]: Nice 👍🏻

Comment (kcgen) [reply tests/bit_view_tests.cpp:33]: Thanks!

It's about as far I think we can go and keep it in a very generic and clean C++ form.

It's now more convenient and safe, in that the `bit_view` will always use the correct `uint*_t` to fully cover the view. 

Compile-time checks guarantee that illegal views (negative index, zero or negative width, and views that overhang the start or end) aren't compiable - so we've got those hard safety bounds too. I've added unit tests to check these.

For the PC Speaker PR, I'll put this to use managing one of the 8-bit keyboard/PC-speaker registers (PPI controller port B register).

@shermp , shout if you see any gotchas; will hold off a bit before merging. Thanks guys :rocket:

**Comment by kcgen** [comment]
It seems GitHub has pushed an update to the macOS runners, however it's trapped my runners in a crash loop:

``` text
Current runner version: '2.291.1'
2022-06-06 07:28:10Z: Listening for Jobs

Runner update in progress, do not shutdown runner.
Downloading 2.292.0 runner
Runner update process finished.

An error occurred: Runner package 'https://github.com/actions/runner/releases/download/v2.292.0/actions-runner-osx-x64-2.292.0.tar.gz' failed after 3 download attempts
Runner listener exit with retryable error, re-launch runner in 5 seconds.

Restarting runner...


� Connected to GitHub

Current runner version: '2.291.1'
2022-06-06 07:28:24Z: Listening for Jobs

Runner update in progress, do not shutdown runner.
Downloading 2.292.0 runner
Runner update process finished.

An error occurred: Runner package 'https://github.com/actions/runner/releases/download/v2.292.0/actions-runner-osx-x64-2.292.0.tar.gz' failed after 3 download attempts

Runner listener exit with retryable error, re-launch runner in 5 seconds.

...
```

File: tests/bit_view_tests.cpp
Comment (johnnovak) [reply tests/bit_view_tests.cpp:33]: Great stuff, I agree, this strikes a good balance between usability/safety and simplicity of implementation.

It will be handy for my upcoming AdLib Gold stuff as well. I got it already working in a rather hacky way and it sounds glorious 😁

Comment (kcgen) [reply tests/bit_view_tests.cpp:33]: > It will be handy for my upcoming AdLib Gold stuff as well. I got it already working in a rather hacky way and it sounds glorious :grin:

:exploding_head: :heart_eyes: .

**Comment by kcgen** [comment]
Thanks for the review @shermp and @johnnovak.

> From an ergonomics point of view, is there a way of creating the bit_view without explicitly setting the type? 

@shermp, your suggestion is now addressed safely and with compile-time asserts; so it's impossible to read or write before or after the view's data value.

Unit tests are passing on all platforms, along with passing all the *SAN tests locally.  

Merging.

**Pull request merged**
